documents/000-base-scope.md

```diff
@@ -0,0 +1,189 @@
+pull_requests: [3]
+published:
+last_modified: 2020-05-06
+---
+
+# Base Scope
+
+These guidelines apply to a syntax definition's base scope,
+which is applied to the entire document
+highlighted using that syntax.
+Additionally, the scope suffix,
+usually based on the base scope,
+is explained.
+
+
+## Areas of Concern
+
+- Base Scope
+- Scope suffix
+
+
+## Rationale
+
+To classify a document type,
+every syntax specifies a base scope name,
+which is applied to the entire document
+highlighted as that syntax.
+The base scope can be used
+to differentiate between, for example,
+(mostly) prose text documents
+and documents containing instructions (*source code*).
+
+Furthermore,
+to identify which syntax a certain token is matched from (or as),
+every scope name should use a semi-unique scope suffix
+that is appended as the last sub-scope.
+
+
+## Guidelines
+
+### Base Scope
+
+The base scope is applied to the entire region
+lexed by a syntax definition as the very first scope.
+Usually that will be the entire file,
+but in the case of an embedded syntax,
+a region might have multiple stacked base scopes.
+In the context of a file, however,
+there will only be a single base scope
+which is that of the syntax used to lex it initially.
+
+The base scope must be a specialization
+of the following three first-level sub-scopes:
+
+- `text` shall be used
+  for syntaxes that contain primarily prose text,
+  such as markup languages
+  and structured data formats
+  with a focus on text.
+  Text in such syntaxes is spell-checked
+  and auto completion is disabled while typing.
+
+- `source` shall be used
+  for programming languages containing
+  instructions for a compiler, an interpreter, or similar.
+  Text in such syntaxes usually has special meaning,
+  is not spell-checked and auto completions are enabled.
+
+- `embedding` special use case for syntaxes
+   that are embedded later into the file.
+   The effect of an `embedding` scope is mostly transparent
+   and the text vs. source decision
+   is deferred to the following scope name.
+
+The second sub-scope level of the scope name
+should be based on the
+Alphanumeric representations should be preferred.
+A notable exception is `source.c++`.
+
+
+#### Specialized Base Scopes
+
+When implementing syntax highlighting
+for a dialect of a different syntax,
+for example a special syntax
+for JSON documents
+or a Bash-like shell scripting language,
+a third-level scope can be introduced
+to indicate that a syntax is a specialization of the base.
+Examples include Markdown being a specialization of HTML
+or Zsh being a specialization of Bash,
+resulting in `text.html.markdown`.
+If the syntax for the base is known to be specialized,
+it may opt to use the `basic` sub-scope
+instead of simply omitting it.
+
+
+### Scope Suffix
+
+A scope suffix is used for individual tokens.
+The suffix is appended to every scope name
+that is assigned to a token,
+but there may only be one such suffix.
```

Comment:
To be clear, this is not referring to uniqueness but that the suffix generally consists of one scope atom (using terminology from syntect here) - I misinterpreted at first until I read further down and thought there was a contradiction before realising.

Reply:
Good point, I'll rephrase that.